### PR TITLE
Full review of map/reduce documentation - Issue 167

### PR DESCRIPTION
**Pull request by jdgarciauc3m**

### CODE REVIEW

File: doc/map-reduce.md [hunk 1]
```diff
@@ -1,69 +1,123 @@
...
+first argument an execution policy.
 
 ~~~{.cpp}
 grppi::map_reduce(exec, other_arguments...);
 ~~~
 
-**Note**: A **map/reduce** could be also expressed by the composition of a **map** and a **reduce**. However, **map/reduce** fuses both stages allowing for extra optimizations.
+**Note**: A **map/reduce** could be also expressed by the composition of a
+**map** and a **reduce**. However, **map/reduce** fuses both stages allowing for
+extra optimizations.
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `stages allowing for extra optimizations` by `stages, thus allowing for further optimizations`.

File: doc/map-reduce.md [hunk 2]
```diff
@@ -1,69 +1,123 @@
...
 
+A unary **map/reduce** also requires an identity value for the **Combiner**.
+
 ---
-**Example**
+**Example**: Transforms a sequence of strings to its corresponding double values
+and computes the addition of those values.
 ~~~{.cpp}
+vector<string> v { "1.0", "2.0", "3.5", "0.25" };
 auto res = map(exec,
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `map` by `grppi::map_reduce`.

File: doc/map-reduce.md [hunk 3]
```diff
@@ -1,69 +1,123 @@
...
 ~~~
 ---
 
 
-### N-ary map_reduce
+### N-ary map/reduce
 
-A n-ary **map/reduce** takes multiple data sets and performs consecutively the **map** and **reduce** stage, returning the reduced value.
+A n-ary **map/reduce** takes multiple data sets and performs consecutively the
+**map** and **reduce** stage, returning the reduced value.
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `stage` by `stages`.

File: doc/map-reduce.md [hunk 4]
```diff
@@ -1,69 +1,123 @@
...
 
-The only interface currently offered for this pattern is based in iterators (following the C++ standard library conventions):
+The only interface currently offered for this pattern is based in iterators
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `based in` by `based on`.

File: doc/map-reduce.md [hunk 5]
```diff
@@ -1,69 +1,123 @@
...
+(following the C++ standard library conventions):
 
   * The first data set is specified by two iterators.
-  * All the other input data sets are specified by iterators to the start of the input data sequences, assuming that the size of all sequences are at least as large as the first sequence.
+  * All the other input data sets are specified by iterators to the start of the
+    input data sequences, assuming that the size of all sequences are at least
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `are at least as` by `are, at least, as`.

File: doc/map-reduce.md [hunk 6]
```diff
@@ -1,69 +1,123 @@
...
+
+A n-ary **map/reduce** also requires an identity value for the **Combiner**.
 
 ---
-**Example**
+**Example**: Compute scalar vector between two vectors of doubles.
 ~~~{.cpp}
+v = get_first_vector();
+w = get_second_vector();
 auto res = map(exec,
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `map` by `grppi::map_reduce`.

File: include/tbb/mapreduce.h [hunk 7]
```diff
@@ -25,10 +25,37 @@
...
+/**
+\addtogroup mapreduce_pattern_tbb TBB parallel map/reduce pattern
+\brief TBB parallel implementation of the \ref md_map-reduce pattern.
+@{
+*/
+
+/**
+\brief Invoke [map/reduce pattern](\ref md_map-reduce) on a data sequence with 
+TBB parallel execution.
+\tparam InputIt Iterator type used for input sequence.
```
Comment (mdolz) [inline include/tbb/mapreduce.h]: Replace `for input` by `for the input`.

File: include/mapreduce.h [hunk 8]
```diff
@@ -21,53 +21,18 @@
...
- *
- *  @brief Apply the mapreduce pattern for parallelizing the code section.
- *
- *	The Mapreduce pattern is based on the application of a function 'map' 
- *	to all the elements of a given data structures 'in', and then merge
- * 	the result to another variable using binary elemental operations.
- *	In a Mapreduce the datastructure is divided in as many data sections as 
- *	threads are available to be used by the function that will execute in
- *	in parallel the code section 'map'. Each task will reduce its own 
- *	data section. Finaly all the thread section will be join in an output
```
Comment (mdolz) [inline include/mapreduce.h]: Replace `Finaly` by `Finally`.
Replace `will be join` by `will be joined`.

File: include/seq/mapreduce.h [hunk 9]
```diff
@@ -22,22 +22,53 @@
...
+/**
+\addtogroup mapreduce_pattern_seq Sequential map/reduce pattern
+\brief Sequential implementation of the \ref md_map-reduce pattern.
+@{
+*/
+
+/**
+\brief Invoke [map/reduce pattern](\ref md_map-reduce) on a data sequence with 
+sequential execution.
+\tparam InputIt Iterator type used for input sequence.
```
Comment (mdolz) [inline include/seq/mapreduce.h]: Replace `for input` by `for the input`.

File: include/poly/mapreduce.h [hunk 10]
```diff
@@ -60,19 +60,52 @@ Identity map_reduce_multi_impl(polymorphic_execution & e, InputIt first, InputIt
...
+/**
+\addtogroup mapreduce_pattern_poly Polymorphic map/reduce pattern
+\brief Polymorphic implementation of the \ref md_map-reduce pattern.
+@{
+*/
+
+/**
+\brief Invoke [map/reduce pattern](\ref md_map-reduce) on a data sequence with 
+polymorphic execution.
+\tparam InputIt Iterator type used for input sequence.
```
Comment (mdolz) [inline include/poly/mapreduce.h]: Replace `for input` by `for the input`.

File: include/native/mapreduce.h [hunk 11]
```diff
@@ -22,47 +22,82 @@
...
+/**
+\addtogroup mapreduce_pattern_native Native parallel map/reduce pattern
+\brief Native parallel implementation of the \ref md_map-reduce pattern.
+@{
+*/
+
+/**
+\brief Invoke [map/reduce pattern](\ref md_map-reduce) on a data sequence with 
+native parallel execution.
+\tparam InputIt Iterator type used for input sequence.
```
Comment (mdolz) [inline include/native/mapreduce.h]: Replace `for input` by `for the input`.

File: doc/map-reduce.md [hunk 12]
```diff
@@ -1,69 +1,123 @@
...
+output type `U`, makes valid the following:
+
+~~~{.cpp}
+U res = op(x);
+~~~
+
+A **MultiTransformer** is any C++ callable entity that takes data items, one of
+each input sequence, and transforms them into an output value. The input types
+and the output type may differ. Thus, a multi-transformer `op` is any operation
+that, given inputs `x1, x2, ... , xN` of types `T1, T2, ... , TN` and an output
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `given inputs` by `given the inputs`.

**Review by mdolz — CHANGES_REQUESTED** [review]
changes requested are inline.

File: doc/map-reduce.md (hunk 1 above)
Comment (jdgarciauc3m) [reply doc/map-reduce.md]: Rephrased to:

However, **map/reduce** may fuse both stages, allowing for extra optimizations.

File: doc/map-reduce.md (hunk 12 above)
Comment (jdgarciauc3m) [reply doc/map-reduce.md]: OK

File: doc/map-reduce.md (hunk 2 above)
Comment (jdgarciauc3m) [reply doc/map-reduce.md]: OK

File: doc/map-reduce.md (hunk 3 above)
Comment (jdgarciauc3m) [reply doc/map-reduce.md]: OK

File: doc/map-reduce.md (hunk 4 above)
Comment (jdgarciauc3m) [reply doc/map-reduce.md]: OK

File: doc/map-reduce.md (hunk 5 above)
Comment (jdgarciauc3m) [reply doc/map-reduce.md]: Good catch!

File: include/tbb/mapreduce.h (hunk 7 above)
Comment (jdgarciauc3m) [reply include/tbb/mapreduce.h]: OK

File: include/seq/mapreduce.h (hunk 9 above)
Comment (jdgarciauc3m) [reply include/seq/mapreduce.h]: OK

File: include/poly/mapreduce.h (hunk 10 above)
Comment (jdgarciauc3m) [reply include/poly/mapreduce.h]: OK

File: include/native/mapreduce.h (hunk 11 above)
Comment (jdgarciauc3m) [reply include/native/mapreduce.h]: OK

File: doc/map-reduce.md [hunk 13]
```diff
@@ -1,69 +1,123 @@
...
+A n-ary **map/reduce** also requires an identity value for the **Combiner**.
 
 ---
-**Example**
+**Example**: Compute scalar vector between two vectors of doubles.
 ~~~{.cpp}
-auto res = map(exec,
+v = get_first_vector();
+w = get_second_vector();
+auto res = map_reduce(exec,
```
Comment (mdolz) [inline doc/map-reduce.md]: Replace `map_reduce` by `grppi::map_reduce`.

**Review by mdolz — CHANGES_REQUESTED** [review]

**Review by Drioastorga — DISMISSED** [review]
Ok

**Review by mdolz — APPROVED** [review]

**Review by javidoc00 — APPROVED** [review]

**Comment by jdgarciauc3m** [comment]
Closes issue #167

**Pull request merged**